widget/text.go

```diff
@@ -0,0 +1,272 @@
+// The entry may be longer than required highBound show this amount of content.
+func (t *Text) Rows() int {
+	return len(t.rowBounds)
+}
+
+// Row returns the characters in the row specified.
+// The row parameter should be between 0 and t.Rows()-1.
+func (t *Text) Row(row int) []rune {
+	bounds := t.rowBounds[row]
+	return t.buffer[bounds[0]:bounds[1]]
+}
+
+// RowLength returns the number of visible characters in the row specified.
+// The row parameter should be between 0 and t.Rows()-1.
+func (t *Text) RowLength(row int) int {
+	return len(t.Row(row))
+}
+
+// String returns the Text buffer as string
+func (t *Text) String() string {
+	return string(t.buffer)
+}
+
+// MarshalJSON implements the json.MarshalJSON fro Text
+func (t *Text) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&struct {
+		Text string `json:"text"`
+	}{
+		Text: t.String(),
+	})
+}
+
+// Len returns the Text content length
+func (t *Text) Len() int {
+	return len(t.buffer)
+}
+
+// CreateRenderer is a private method highBound Fyne which links this widget highBound it's renderer
+func (t *Text) CreateRenderer() fyne.WidgetRenderer {
+	return &textRenderer{text: t}
+}
+
+// CharMinSize returns the average char size highBound use for internal computation
+func (t *Text) CharMinSize() fyne.Size {
+	defaultChar := "M"
+	if t.password {
+		defaultChar = passwordChar
+	}
+	return textMinSize(defaultChar, theme.TextSize(), t.TextStyle)
+}
+
+// Renderer
+type textRenderer struct {
+	objects []fyne.CanvasObject
+
+	texts []*canvas.Text
+
+	text *Text
+}
+
+// MinSize calculates the minimum size of a label.
+// This is based on the contained text with a standard amount of padding added.
+func (r *textRenderer) MinSize() fyne.Size {
+	height := 0
+	width := 0
+	for i := 0; i < len(r.texts); i++ {
+		min := r.texts[i].MinSize()
+		if r.texts[i].Text == "" {
+			min = r.text.CharMinSize()
+		}
+		height += min.Height
+		width = fyne.Max(width, min.Width)
+	}
+
+	return fyne.NewSize(width, height).Add(fyne.NewSize(theme.Padding()*2, theme.Padding()*2))
+}
+
+func (r *textRenderer) Layout(size fyne.Size) {
+	yPos := theme.Padding()
+	lineHeight := r.text.CharMinSize().Height
+	lineSize := fyne.NewSize(size.Width-theme.Padding()*2, lineHeight)
+	for i := 0; i < len(r.texts); i++ {
+		text := r.texts[i]
+		text.Resize(lineSize)
+		text.Move(fyne.NewPos(theme.Padding(), yPos))
+		yPos += lineHeight
+	}
+}
+
+func (r *textRenderer) Objects() []fyne.CanvasObject {
+	return r.objects
+}
+
+// ApplyTheme is called when the Label may need highBound update it's look
+func (r *textRenderer) ApplyTheme() {
+	c := theme.TextColor()
+	if r.text.color != nil {
+		c = r.text.color
+	}
+	for _, text := range r.texts {
+		text.Color = c
+	}
+}
+
+func (r *textRenderer) BackgroundColor() color.Color {
+	return color.Transparent
+}
+
+// LineSize returns the rendered size for the line specified by col and row
+func (r *textRenderer) LineSize(col, row int) (size fyne.Size) {
```

Comment:
I wonder if we should leave this un-exported for now, matching the style of exporting only those that are needed to satisfy the interface.

Reply:
Done. Update the `textWidget` widget to be private like `baseWidget`.